### PR TITLE
fix: improve startup times by initializing projection tables during setup and reduce initial load

### LINKED ISSUE #3588: Reduce startup times
For being able to scale more reactively, reduced startup times could help.
One way of achieving that could be to separate init and start phases.

We do have 3 different parts right now: start, setup and init
The start should not check the projections

**Acceptance Criteria**
- [ ] Start should not check projections
- [ ] Spooler should not check for all instances if they are done, just check last cycle

### PR DESCRIPTION
**Pull request by livio-a**
closes #3588

### CODE REVIEW

File: internal/eventstore/handler/crdb/init.go [hunk 1]
```diff
@@ -186,35 +186,30 @@ type ForeignKey struct {
...
-		if err := tx.Commit(); err != nil {
+			tx.Rollback()
 			return err
 		}
+		if !next {
+			logging.WithFields("projection", h.ProjectionName, "execute", i).Debug("skipping next check")
+			break
+		}
+	}
+	if err := tx.Commit(); err != nil {
```
Comment (adlerhurst) [inline internal/eventstore/handler/crdb/init.go:210]: Simplify to return tx.Commit()

File: internal/query/projection/projection.go [hunk 2]
```diff
@@ -63,7 +62,16 @@ var (
...
 )
 
-func Start(ctx context.Context, sqlClient *sql.DB, es *eventstore.Eventstore, config Config, keyEncryptionAlgorithm crypto.EncryptionAlgorithm, certEncryptionAlgorithm crypto.EncryptionAlgorithm) error {
+type projection interface {
+	Start()
+	Init(ctx context.Context) error
+}
+
+var (
+	projections = make([]projection, 42)
```
Comment (adlerhurst) [inline internal/query/projection/projection.go:71]: This memory allocation is useless imo

File: internal/eventstore/handler/crdb/init.go [hunk 3]
```diff
@@ -186,36 +186,28 @@ type ForeignKey struct {
...
-				logging.WithFields("projection", h.ProjectionName, "execute", i).Debug("skipping next check")
-				break
-			}
-		}
-		if err := tx.Commit(); err != nil {
+			tx.Rollback()
 			return err
 		}
+		if !next {
+			logging.WithFields("projection", h.ProjectionName, "execute", i).Debug("skipping next check")
```
Comment (adlerhurst) [inline internal/eventstore/handler/crdb/init.go:206]: Maybe "projection set up" would be more clear than skip?

File: internal/eventstore/handler/handler_projection.go [hunk 4]
```diff
@@ -209,32 +239,82 @@ func (h *ProjectionHandler) schedule(ctx context.Context) {
...
+		AddQuery().
+		AggregateTypes(aggregateType).
+		AggregateIDs(aggregateID).
+		EventTypes(schedulerSucceeded).
+		EventData(map[string]interface{}{
+			"name": h.ProjectionName,
+		}).
+		Builder(),
+	)
+	if err != nil {
```
Comment (adlerhurst) [inline internal/eventstore/handler/handler_projection.go:289]: Simplify return by len(...) > 0 && err != nil, err

Comment (adlerhurst) [reply internal/eventstore/handler/handler_projection.go:289]: And lines below

File: internal/eventstore/handler/handler_projection.go [hunk 5]
```diff
@@ -209,32 +239,82 @@ func (h *ProjectionHandler) schedule(ctx context.Context) {
...
+		return false, err
+	}
+	return len(events) > 0, nil
+}
+func (h *ProjectionHandler) setSucceededOnce(ctx context.Context) error {
```
Comment (adlerhurst) [inline internal/eventstore/handler/handler_projection.go:294]: Add new line before (linting)

File: internal/eventstore/handler/crdb/init.go (hunk 3 above)
Comment (livio-a) [reply internal/eventstore/handler/crdb/init.go:206]: a projection can have multiple checks, e.g. index check after table check, ...
this is unchanged BTW

Comment (livio-a) [reply internal/eventstore/handler/crdb/init.go:206]: oh yes you're right. i'll change it

File: internal/eventstore/handler/handler_projection.go (hunk 4 above)
Comment (livio-a) [reply internal/eventstore/handler/handler_projection.go]: `...&& err == nil...` is even more helpful 😉

**Review by adlerhurst — APPROVED** [review]

**Pull request merged**